File: pkg/graph/label_test.go
package graph

import (
	"strings"
	"testing"
)

// NormalizeLabel mirrors normalizeLabel() in src/graph/label.ts, and
// the two must never drift: a label the editor accepts has to survive
// an MCP round-trip byte-identically, or agents and humans see
// different text. The cases here are the TS suite's, ported — if one
// side gains a case, port it to the other.
func TestNormalizeLabelMirrorsTheEditor(t *testing.T) {
	cases := []struct {
		name string
		in   string
		want string
	}{
		{"empty stays empty", "", ""},
		{"trims leading and trailing whitespace", "   hello   ", "hello"},
		{"collapses internal runs to one space", "a   b\tc", "a b c"},
		{"preserves explicit newlines", "first\nsecond", "first\nsecond"},
		{"trims per line, keeps interior blank lines", "  a  \n\n  b  ", "a\n\nb"},
		{"drops fully-blank leading and trailing lines", "\n\nhi\n\n", "hi"},
		{"whitespace-only collapses to empty", " \t \n \t ", ""},
		{"CRLF and CR both normalize to LF", "a\r\nb\rc", "a\nb\nc"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := NormalizeLabel(c.in); got != c.want {
				t.Fatalf("NormalizeLabel(%q) = %q, want %q", c.in, got, c.want)
			}
		})
	}
}

func TestNormalizeLabelCapsOnRuneBoundary(t *testing.T) {
	// The cap must count runes, not bytes: a multi-byte codepoint
	// sliced in half is invalid UTF-8, which the line-based .flowgo
	// serializer would then write out corrupted.
	long := strings.Repeat("ü", MaxLabelLen+50)
	got := NormalizeLabel(long)
	if n := len([]rune(got)); n != MaxLabelLen {
		t.Fatalf("capped length = %d runes, want %d", n, MaxLabelLen)
	}
	if !strings.HasSuffix(got, "ü") {
		t.Fatalf("cap split a multi-byte rune: label ends in %q", got[len(got)-4:])
	}

	exact := strings.Repeat("x", MaxLabelLen)
	if got := NormalizeLabel(exact); got != exact {
		t.Fatalf("a label exactly at the cap must pass through unchanged")
	}
}

File: src/graph/label.test.ts
import { describe, expect, it } from "vitest";
import { MAX_LABEL_LEN, normalizeLabel } from "./label";

describe("normalizeLabel", () => {
  it("returns empty + not-truncated for null / undefined / empty input", () => {
    expect(normalizeLabel(null)).toEqual({ label: "", truncated: false });
    expect(normalizeLabel(undefined)).toEqual({ label: "", truncated: false });
    expect(normalizeLabel("")).toEqual({ label: "", truncated: false });
  });

  it("trims leading and trailing whitespace", () => {
    expect(normalizeLabel("   hello   ")).toEqual({
      label: "hello",
      truncated: false,
    });
  });

  it("collapses internal non-newline whitespace runs to a single space", () => {
    expect(normalizeLabel("a   b\tc")).toEqual({
      label: "a b c",
      truncated: false,
    });
  });

  it("preserves explicit newlines as hard line breaks", () => {
    expect(normalizeLabel("first\nsecond")).toEqual({
      label: "first\nsecond",
      truncated: false,
    });
  });

  it("trims per-line and keeps interior blank lines", () => {
    expect(normalizeLabel("  a  \n\n  b  ")).toEqual({
      label: "a\n\nb",
      truncated: false,
    });
  });

  it("drops fully-blank leading and trailing lines", () => {
    expect(normalizeLabel("\n\nhi\n\n")).toEqual({
      label: "hi",
      truncated: false,
    });
  });

  it("normalises CRLF and CR to LF", () => {
    expect(normalizeLabel("a\r\nb\rc")).toEqual({
      label: "a\nb\nc",
      truncated: false,
    });
  });

  it("preserves single internal spaces", () => {
    expect(normalizeLabel("Should we roll a dice?")).toEqual({
      label: "Should we roll a dice?",
      truncated: false,
    });
  });

  it("does not strip punctuation, quotes, parens, or non-ASCII", () => {
    const sample = '/fmt.Errorf("line %d: edge needs from to", lineNo)?';
    expect(normalizeLabel(sample)).toEqual({
      label: sample,
      truncated: false,
    });
    expect(normalizeLabel("résumé — “smart quotes”").label).toBe(
      "résumé — “smart quotes”",
    );
  });

  it("hard-caps at MAX_LABEL_LEN by default and reports truncation", () => {
    const long = "x".repeat(MAX_LABEL_LEN + 50);
    const result = normalizeLabel(long);
    expect(result.label.length).toBe(MAX_LABEL_LEN);
    expect(result.truncated).toBe(true);
  });

  it("respects a caller-supplied cap", () => {
    expect(normalizeLabel("abcdef", { maxLength: 3 })).toEqual({
      label: "abc",
      truncated: true,
    });
  });

  it("does not flag truncation when within the cap", () => {
    expect(normalizeLabel("under cap").truncated).toBe(false);
  });

  it("preserves multi-line code-shaped paste with per-line trimming", () => {
    const multi = "func main() {\n    fmt.Println(\"hi\")\n}";
    expect(normalizeLabel(multi).label).toBe(
      'func main() {\nfmt.Println("hi")\n}',
    );
  });

  // Astral-plane characters (many emoji, e.g. U+1F600 😀) are two
  // UTF-16 code units but one Unicode codepoint. Capping by
  // `string.length` (UTF-16 units) hits MAX_LABEL_LEN at half the
  // intended visible-character count, and `.slice(0, cap)` can split
  // a surrogate pair in half, leaving an unpaired surrogate in the
  // output. This must count and truncate by codepoint, matching
  // pkg/graph.NormalizeLabel's `[]rune` cap — the two share one
  // MAX_LABEL_LEN and must agree on what it counts.
  it("caps by Unicode codepoint, not UTF-16 code unit", () => {
    const emoji = "😀".repeat(MAX_LABEL_LEN + 10); // each is 2 UTF-16 units
    const result = normalizeLabel(emoji);
    expect(result.truncated).toBe(true);
    expect(Array.from(result.label).length).toBe(MAX_LABEL_LEN);
    // No unpaired surrogate: the string must still be valid UTF-16,
    // i.e. re-splitting into codepoints and joining reproduces it.
    expect(Array.from(result.label).join("")).toBe(result.label);
  });

  it("does not truncate a label with exactly MAX_LABEL_LEN codepoints of emoji", () => {
    const emoji = "😀".repeat(MAX_LABEL_LEN);
    const result = normalizeLabel(emoji);
    expect(result.truncated).toBe(false);
    expect(Array.from(result.label).length).toBe(MAX_LABEL_LEN);
  });
});
